### ISSUE TITLE
Ensure that existing manipulation and analysis functions use the `mask` where appropriate.

### ISSUE THREAD

**Issue by eteq** [astropy/specutils#516]
This comes out of #509, which introduces a function that computes and assigns `mask`'s to `Spectrum1D` objects.  This led to a discussion where it was realized that a lot of the downstream analysis tasks that would follow on from #509 need to respect the `mask` for #509 to actually be practically useful - without that it seemed like a `flux_masked` or similar seemed necessary, which is not desirable since it essentially replicates the mask in `Spectrum1D` itself.

So the point of this issue is to make sure the currently-existing `analysis` and `manipulation` tools are mask-aware. At the time of writing this, I think the current status is:

* `specutils.analysis` functions: `centroid`, `equivalent_width`, `fwhm`, `gaussian_fwhm`, `gaussian_sigma_width`, `line_flux`, `snr` - all of these ignore the mask entirely, but I think it's a pretty easy thing to make them work: if the mask is `None`, do what it does now, otherwise, just do `self.flux[~self.mask]` where `self.flux` currently appears (and `self.spectral_axis` -> `self.spectral_axis[~self.mask]`, although they don't all use the spectral axis).
* line and continuum fitting (same underlying implementation. in `fit_lines`): I think this is *also* basically a one-liner in the implementation, doing essentially the above, but in the call to the fitter.
* parameter estimators in @brechmos-stsci may be able to advise here, but I *think* these are all eventually implemented via the `analysis` functions, right?  So then if we fix those then it automatically carries over to the estimators? 
* `extract_region`: already does the correct thing of just carrying through the mask.

The one question in my mind that applies to all of the above: do we want `use_mask` to be a keyword we add to *all* of these?  My initial inclination is no: we always use the mask, but if the user does not want that, they can just set the mask to None. Or alternatively, 

cc @brechmos-stsci @camipacifici @nmearl

**Comment by eteq** [comment]
Oh and since I already looked at it, I'll make a PR for `fit_lines`.

**Comment by hcferguson** [comment]
This looks sensible to me.

Regarding without_mask, it seems like there should be some way of doing this without copying any data. In the case where the user substitutes the mask with None, there is no copy. But it's inconvenient for the user to get the mask back. It would be nice to be able to toggle the mask on and off, which is what you are getting at, but do it by just temporarily changing the reference of the mask attribute. 

If it were possible, it might be simpler for there to be spectrum.unmasked as a version of the *same* spectrum1D instance, but with the mask set to None. I couldn't instantly think of a way to implement this though. So calls would be `function(spectrum.unmasked)` rather than `function(spectrum.without_mask())`.

**Comment by eteq** [comment]
> If it were possible, it might be simpler for there to be spectrum.unmasked as a version of the same spectrum1D instance,

Hmm, one option would be to do a *shallow* copy - e.g., it would be a copy of the `Spectrum1D` object, but the `flux` (and `spectral_axis`, `uncertainty`, etc) objects would still be referring to the same arrays.  Then it's still a copy but a fairly "lightweight" copy.  Do you think that would work, @hcferguson ?

**Comment by hcferguson** [comment]
Might work. But I think that `X = spectrum.whatever()` makes you think you are getting a deep copy. Whereas `X = spectrum.whatever` makes it clearer that you are just getting a sort of a view. But I can't think of another example where an attribute basically refers to a different version of the instance itself -- with one of its other attributes modified -- so this may be equally confusing.

**Comment by jdavies-st** [comment]
Does #670 resolve this issue @ibusko?  If so, this issue can be closed.

**Comment by rosteen** [comment]
@jdavies-st @ibusko I'm still seeing `line_flux` and `equivalent_width` returning `NaN [unit]` on masked arrays after #670, can someone else test and see if they're getting the same result? I assume they were working to merge that PR, but I haven't had time to track down exactly why it's failing. `centroid`, `gaussian_fwhm` and `gaussian_sigma` all work in my testing (I'm using these in a line analysis plugin in specviz).

**Comment by nmearl** [comment]
@rosteen can you provide an example of what you're doing? I'm playing around with the analysis functions, but am not getting  any `NaN` values returned.

**Comment by rosteen** [comment]
I'm using the `specviz` UI to define a subset around a line, then using `data = specviz.app.get_data_from_viewer("spectrum-viewer", 'Subset 1')` to retrieve the data via the `glue-astronomy` data translators (which I've updated locally to return the mask values that specutils expects). Then I'm just calling `specutils.analysis.line_flux(data["Subset 1"])`

I attached zipped numpy save files for the flux, spectral_axis and mask for replication. Load them into numpy as flux, spec, and mask, and then run the following to replicate:

```
flux = flux*u.Unit('1e-17 erg / (Angstrom cm2 s)')
spec = spec*u.AA
spec1d = Spectrum1D(flux=flux, spectral_axis=spec, mask=mask)
analysis.line_flux(spec1d)
```

[mask_test.zip](https://github.com/astropy/specutils/files/4949210/mask_test.zip)

**Comment by nmearl** [comment]
The flux array is all `NaN`s. Looking at the returned value of `data` when trying to call `get_data_from_viewer` on a subset shows that the flux value is `NaN`s even there. Perhaps something is off in the translator when dealing with subsets.

That said, subsets _are_ generalized concepts, unlike data. It might be more consistent to use the dedicated spectral region methods in this case:

<img width="790" alt="Screen Shot 2020-07-20 at 1 56 23 PM" src="https://user-images.githubusercontent.com/4141126/87969836-da9c5000-ca90-11ea-9d85-b744465c397c.png">

**Comment by rosteen** [comment]
The flux array is not all NaNs: 
```
>>> flux[np.where(~mask)]
<Quantity [ 58.38358 ,  56.5681  ,  57.41874 ,  56.701534,  57.98195 ,
(...)
52.95215 ,  52.214996,  53.889793,  55.43877 ,  54.863873] 1e-17 erg / (Angstrom cm2 s)>

>>> flux[np.where(~mask)].shape
(255,)
```

**Comment by nmearl** [comment]
They are. Both the serialized files you gave and when I attempt to retrieve the subset data through the `get_data_from_viewer` method returns `NaN` values:

<img width="735" alt="Screen Shot 2020-07-20 at 2 02 52 PM" src="https://user-images.githubusercontent.com/4141126/87970488-d3297680-ca91-11ea-9378-e51f9fa854db.png">

<img width="1090" alt="Screen Shot 2020-07-20 at 2 03 08 PM" src="https://user-images.githubusercontent.com/4141126/87970543-e4728300-ca91-11ea-84c8-132a38cd4dbf.png">

**Comment by rosteen** [comment]
But your code snippet doesn't show that all the values are NaN, it shows that the first and last three values are NaN (which is expected, since the selected subset is in the middle of the spectrum). Did you get NaN trying either of the commands I posted above, which should be returning real values? (e.g. `np.nanmax(flux)`)?

**Comment by nmearl** [comment]
Right, (I deleted the comment almost as soon as I posted). The problem now is that this isn't an issue with specutils; we should change the glue-astronomy translator to not set the values [outside the subset to `nan`](https://github.com/glue-viz/glue-astronomy/blob/master/glue_astronomy/translators/spectrum1d.py#L94).

The alternative would be as you pointed out: using `nan` function values, but we had been trying to avoid using functions that make implicit assumptions of users' data (i.e. implicitly handling `nan` values can returned unintended results to the user).

**Comment by rosteen** [comment]
Haha, sorry - I was actively looking at/into this and caught your comment too quickly. 

I would argue (and this is getting into what the acceptance criteria for closing this issue are, so maybe we need @eteq or @hcferguson to chime in again) that this is definitely still a specutils issue, because as a user I would be rather dismayed if the values I had masked out (be they NaNs or anything else) were having an effect on these calculations. Especially since the NaNs would be ignored simply by doing something like applying the calculations to `flux[~mask]` (which is exactly what @eteq recommended in the initial description of this issue). We really (in my opinion) shouldn't need to be changing the regions outside the mask to non-NaN or using NaN-handling functions, since data points outside the mask shouldn't even be visible to the calculations.

**Comment by nmearl** [comment]
In many cases, such as those that require information about the the shape of the data, or that bins be consistent, we can't rely on just removing pieces of a spectrum. See Erik's comment [here](https://github.com/astropy/specutils/pull/670#pullrequestreview-408372752) in which he highlights the implementation that `line_flux` uses: masked values are not ignored, but interpolated over, because we need shape consistency and reliable `dx` values.

It may be that we're at a crossroads such that we need to spend time implementing option 3, but currently,

> since data points that have been masked out shouldn't even be visible to the calculations.

is not strictly true. There exist cases where we need to rely on uniform binning that simply removing chunks of the spectrum from calculations prevents us from ensuring.
